### ISSUE TITLE
Add auto-exposure and threshold setting

### ISSUE THREAD

**Issue by TheMariday** [TheMariday/marimapper#33]

**Comment by tschundler** [comment]
This is how my tool works. Then there's no problem with something bright-ish in the frame confusing it and less need to fuss with threshold values. So I grab several "dark" frames, average and do a blur on it.

Thinking about it again now, rather than a normal blur, I'd take the highest value within a radius. It would end up looking like camera bokeh. That would allow for some minor camera movement / minor background movement.

To count as "not dark" I know I experimented with dark  vs average brightness across time (assuming off at least 50% of the time) vs lowest 20-percentile + standard deviation of the value or something like that. If you're interested, I could check exactly what the latest code is doing. (but that doesn't mean it is the best)

---

it gets to be a lot of hyper-parameters quickly. Another thing I considered was using some ML type tools to explore the hyperparameters processing the same dataset different ways to find a good combination. (since some things like how I do thresholding could also impact whether boom filter matching gives better results or worse results)

---

I've thought if I ever build a GUI, ideally I'd also let the user manually mask out tricky areas (eg places with reflections) if they want more accuracy.

**Comment by TheMariday** [comment]
@simap I'm sorry you couldn't get any webcams working on Mac. Sadly I don't own a mac so can't do a huge amount of testing compared to windows or linux.

And yes bright mode is something I'm thinking about at the moment. Can be addressed once #50 is done.

Thank you for the feedback and I'll keep you posted with any updates to bright mode

**Comment by JohnnyMarnell** [comment]
Also had trouble on a Mac with multiple webcams. Using my iPhone with free version of Camo Studio app worked pretty well

**Comment by TheMariday** [comment]
Blimey that's some creative thinking!

I'm going to take off the mac support badge for now until I can verify why mac's are having issues with webcams.

I assume the feed appears fine but the exposure controls are missing?

I will look into how opencv handles webcams on mac in the mean time

**Comment by simap** [comment]
> I assume the feed appears fine but the exposure controls are missing?

Yep. Get this in the terminal:

```
[INFO/MainProcess] Connecting to device 0 ...
2024-12-17 13:55:15.965 python[90911:3780639] WARNING: AVCaptureDeviceTypeExternal is deprecated for Continuity Cameras. Please use AVCaptureDeviceTypeContinuityCamera and add NSCameraUseContinuityCameraDeviceType to your Info.plist.
[DEBUG/MainProcess] Connected to device 0 with capture method 0
[INFO/MainProcess] setting cam to dark mode
[DEBUG/MainProcess] Setting autofocus to mode 0 with focus 0
[ERROR/MainProcess] Failed to set autofocus to 0
[ERROR/MainProcess] Failed to set focus to 0
[DEBUG/MainProcess] Setting exposure to mode 0
[ERROR/MainProcess] Failed to put camera into manual exposure mode 0
[DEBUG/MainProcess] Setting gain to 0
[ERROR/MainProcess] failed to set camera gain to 0
[DEBUG/MainProcess] Setting exposure to -10
[ERROR/MainProcess] Failed to set exposure to -10
```

FWIW, this is with the logitech, not the continuity camera. I get similar failed to set xxx messages for the built in camera too. 

The image appears, but its full brightness and it fails to detect anything unless I use webcam settings to dial it down to pretty much a black background. FWIW, I hadn't yet installed webcam settings when I first tried this, so I don't think its a conflict with that app. I also don't have the logitech camera settings app installed on this machine, but that could probably have worked too.

I'm super pleased with how things came out in the end, and didn't spend too much time fiddling with it to get it to work. In the end it saved time over very meticulously placing LEDs and mapping with math and fine tuning by hand. 

The automatic camera and coordinate detection stuff is just awesome.

**Comment by TheMariday** [comment]
@simap @JohnnyMarnell  I've opened a new ticket here #51  but it looks like opencv just doesn't support exposure adjustment on Mac. I'll update the docs to reflect this.

Glad that everything else seems to be working as expected though!

**Comment by tschundler** [comment]
Trying to use this today with my laptop's built-in webcam that only supports auto-exposure was pretty challenging, even at threshold 250 and a dim room

**Comment by TheMariday** [comment]
I think the best approach here is to capture a frame prior to the scan starting and subtract it from subsequent frames.

Just to clarify that marimapper currently doesn't do any dynamic threshold setting

**Comment by russss** [comment]
> I think the best approach here is to capture a frame prior to the scan starting and subtract it from subsequent frames as you have done previously.

I just tried hacking this in and initial impressions are pretty promising using my shitty macbook pro webcam (which obviously doesn't support exposure adjustment). It might be possible to lose the exposure controls entirely.

I'll try and sort out a draft PR with this in but my working copy is a bit of a disaster currently

**Comment by russss** [comment]
I've put an example of my dark-frame subtraction code in https://github.com/TheMariday/marimapper/pull/63 - it's pretty trivial to implement.

This managed to produce the following results using only a Macbook Pro webcam with no exposure adjustments:

![image](https://github.com/user-attachments/assets/4fd4bcda-d7e4-4d7e-903d-8f602c4d3cbb)
![image](https://github.com/user-attachments/assets/d7e1fe7e-9d71-4bf0-80cc-b567fce30bc8)

**Comment by TheMariday** [comment]
Renaming this issue as it's become a discussion of what I will now refer to as "light mode"
